### ISSUE TITLE
AppImage fails to run on Ubuntu 16.04

### ISSUE THREAD

**Issue by probonopd** [sudo-give-me-coffee/flatpak2appdir#3]
Somehow the AppImage still tries to load stuff from the outside, e.g., `/usr/lib/x86_64-linux-gnu/gio/modules/libgvfsdbus.so`, which can fail.

We need to make absolutely sure that the AppImage does not load libraries from the outside.

Can bundling bubblewrap help us there? (Possibly not, because CLONE_NEWNS needs root rights?)

Possibly we need to patch libc/ld-linux to make sure it absolutely never loads anything from outside the AppImage.

This is the error:

```
me@host:~$ /isodevice/Applications/Flatpak-Gedit-1-revised-on-2020-03-18T08-57-43.0-x86_64.AppImage 
/tmp/.mount_FlatpaA9iyme/usr/lib/x86_64-linux-gnu/gio/modules/libgvfsdbus.so: undefined symbol: gvfs_is_ipv6
Failed to load module: /usr/lib/x86_64-linux-gnu/gio/modules/libgvfsdbus.so

(gedit:4365): Gtk-WARNING **: 09:27:11.396: Locale not supported by C library.
	Using the fallback 'C' locale.

(gedit:4365): GLib-GIO-ERROR **: 09:27:11.445: Settings schema 'com.canonical.unity-gtk-module' is not installed
Trace/breakpoint trap

me@host:~$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04 LTS (Xenial Xerus)"
(...)
```

The problem in this case is that `usr/lib/x86_64-linux-gnu/gio/modules/libgvfsdbus.so matches` is in a location that is not on the LD_LIBRARY path nor set as an rpath in all the libraries/binaries that might use it. There may be many other similar locations in which ld files can hide.

I think that systematically rewriting all rpaths might possibly be a way to address this. https://github.com/probonopd/go-appimage/tree/master/src/appimagetool is taking this approach for `-s deploy`.

**Comment by sudo-give-me-coffee** [comment]
By removing `:"${LD_LIBRARY_PATH}"` from these line solved for me, but is really safe to do this?

https://github.com/sudo-give-me-coffee/flatpak2appdir/blob/3182bfc4673d27905d1b8a153459fcb653efa27d/start-app.sh#L26

**Comment by probonopd** [comment]
> By removing :"${LD_LIBRARY_PATH}" from these line works

Does not seem to make a difference on Ubuntu 16.04.

As far as I can see it, we are not telling the linker anywhere that it should load libraries from `<AppDir>/usr/lib/x86_64-linux-gnu/gio/modules/`, hence it loads it from `/usr/lib/x86_64-linux-gnu/gio/modules/`, which fails if that one is "too old".

With https://github.com/probonopd/go-appimage/tree/master/src/appimagetool in `-s deploy` mode, the following test must not return anything:

```
LD_LIBRARY_PATH='' find ./AppDir -type f -exec ldd {} 2>&1 \; | grep '=>' | grep -v AppDir
(returns nothing)
```

In other words, an AppDir that does not depend on shared libraries outside of the AppDir.

I think we need to make it so that flatpak2appdir-generated AppDis also can pass the same test (return nothing).

Does this make sense to you?

**Comment by sudo-give-me-coffee** [comment]
When i do a `strace` i get:
```
access("/home/natanael/Desktop/flatpak2appdir-master/org.gnome.gedit.AppDir/usr/lib/x86_64-linux-gnu/gio/modules/libgvfsdbus.so", F_OK) = 0
```
During runtime everything is loaded under AppDir, anything except themes and icons is loaded outside

**Comment by sudo-give-me-coffee** [comment]
@probonopd can you try this?
https://github.com/sudo-give-me-coffee/flatpak2appdir/releases/download/test/Text_Editor-x86_64.AppImage
I made on Fedora 31 and successfully ran on lubuntu 16.04

**Comment by probonopd** [comment]
When I add a line above the final line in `AppRun` so that it looks like this:

```
find . -type f -exec ldd {} 2>&1 \; | grep '=>' | grep -v AppDir
# exec strace "${HERE}/lib64/ld-linux-x86-64.so.2" --inhibit-cache --library-path "${LIBRARY_PATH}" ${MAIN} "$@"
```

and then  run it, I can see that `ldd` still resolves a lot of stuff to the outside.

**Comment by sudo-give-me-coffee** [comment]
Can you test with latest commit?

---------------------

The AppImage that i found on your fork is missing a lot of libs, this includes some important libs like pixbuf loaders:

```
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xbm.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-ani.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xpm.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-gif.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-pnm.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-png.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-tiff.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-qtif.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-icns.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-bmp.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-tga.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-ico.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-jpeg.so
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.la
< ./lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
```

**Comment by sudo-give-me-coffee** [comment]
> And then run it, I can see that ldd still resolves a lot of stuff to the outside.

LDD is really loading outside, but in strace they not load anything, and building with Travis with all recent commits runs successfully on Ubuntu 16.04 live and even 14.04 on chroot

In my last test i done some rollback and test on Ubuntu 16.04 Live:
* https://github.com/sudo-give-me-coffee/flatpak2appdir/commit/e39bd0538234a15ff0a0d37d4cdb2433951b218c: bundles `gdk-pixbuf-2.0` but still loading gio modules from host so still crashing. 
* https://github.com/sudo-give-me-coffee/flatpak2appdir/commit/6745d51ffaa3eeeac653cb583b1f57352de137ab don't load gio modules from outside and run normaly

**Comment by sudo-give-me-coffee** [comment]
> Possibly we need to patch libc/ld-linux to make sure it absolutely never loads anything from outside the AppImage.

I tried this, `com.github.maoschanz.drawing` and other python apps crashes

> Can bundling bubblewrap help us there? (Possibly not, because CLONE_NEWNS needs root rights?)

Yep

> Does this make sense to you?

Yes, but all Flatpak runtime based apps hash hardcoded paths to these files a rpath doesn't help in this point

**Comment by sudo-give-me-coffee** [comment]
Have you tried this? 
https://github.com/sudo-give-me-coffee/flatpak2appdir/releases/download/test/Text_Editor-x86_64.AppImage

**Comment by probonopd** [comment]
> Have you tried this?
https://github.com/sudo-give-me-coffee/flatpak2appdir/releases/download/test/Text_Editor-x86_64.AppImage

Yes, on Ubuntu 16.04 it results in:

```
access("/tmp/.mount_Text_Ef78DLY/.flatpak-info", F_OK) = -1 ENOENT (No such file or directory)
access("/.flatpak-info", F_OK)          = -1 ENOENT (No such file or directory)
getpid()                                = 8015
poll([{fd=10, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=10, revents=POLLOUT}])
writev(10, [{"\22\0\10\0\1\0\240\3\"\0\0\0\37\0\0\0\10\30\0\0\6\0\0\0gedit\0\0\0"..., 216}, {NULL, 0}, {"", 0}], 3) = 216
poll([{fd=10, events=POLLIN}], 1, -1)   = 1 ([{fd=10, revents=POLLIN}])
recvmsg(10, {msg_name(0)=NULL, msg_iov(1)=[{"\34\0\201\0\1\0\240\3\"\0\0\0s\300\2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 4096}], msg_controllen=0, msg_flags=0}, 0) = 192
poll([{fd=10, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=10, revents=POLLOUT}])
writev(10, [{"b\0\3\0\4\0\240\3", 8}, {"SYNC", 4}, {"", 0}], 3) = 12
poll([{fd=10, events=POLLIN}], 1, -1)   = 1 ([{fd=10, revents=POLLIN}])
recvmsg(10, {msg_name(0)=NULL, msg_iov(1)=[{"\1\0\211\0\0\0\0\0\1\206S\206\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 4096}], msg_controllen=0, msg_flags=0}, 0) = 32
poll([{fd=10, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=10, revents=POLLOUT}])
writev(10, [{"\206\0\2\0\3\1\240\3", 8}, {NULL, 0}, {"", 0}], 3) = 8
poll([{fd=10, events=POLLIN}], 1, -1)   = 1 ([{fd=10, revents=POLLIN}])
recvmsg(10, {msg_name(0)=NULL, msg_iov(1)=[{"\1\0\212\0\0\0\0\0\3\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 4096}], msg_controllen=0, msg_flags=0}, 0) = 32
poll([{fd=10, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=10, revents=POLLOUT}])
writev(10, [{"b\0\3\0\4\0\240\3", 8}, {"SYNC", 4}, {"", 0}], 3) = 12
poll([{fd=10, events=POLLIN}], 1, -1)   = 1 ([{fd=10, revents=POLLIN}])
recvmsg(10, {msg_name(0)=NULL, msg_iov(1)=[{"\1\0\213\0\0\0\0\0\1\206S\206\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 4096}], msg_controllen=0, msg_flags=0}, 0) = 32
poll([{fd=10, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=10, revents=POLLOUT}])
writev(10, [{"\212\2\4\0\1\0\240\0037\1\0\0\7\0\0\0\27\30\2\0007\1\0\0", 24}, {NULL, 0}, {"", 0}], 3) = 24
poll([{fd=10, events=POLLIN}], 1, -1)   = 1 ([{fd=10, revents=POLLIN}])
recvmsg(10, {msg_name(0)=NULL, msg_iov(1)=[{"\1\0\215\0\0\0\0\0\17\0@\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 4096}], msg_controllen=0, msg_flags=0}, 0) = 32
futex(0x7f300c7caf38, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7f300c7caf38, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7f300c7caf38, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7f300c7caf38, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7f300c7caf38, FUTEX_WAKE_PRIVATE, 2147483647) = 0
openat(AT_FDCWD, "/var/lib/snapd/desktop/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = -1 ENOENT (No such file or directory)
access("/tmp/.mount_Text_Ef78DLY/usr/share/glib-2.0/schemas/gschemas.compiled", F_OK) = 0
openat(AT_FDCWD, "/tmp/.mount_Text_Ef78DLY/usr/share/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = 11
fstat(11, {st_mode=S_IFREG|0777, st_size=11464, ...}) = 0
mmap(NULL, 11464, PROT_READ, MAP_PRIVATE, 11, 0) = 0x7f300a0b8000
close(11)                               = 0
access("/tmp/.mount_Text_Ef78DLY/usr/local/share/glib-2.0/schemas/gschemas.compiled", F_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/local/share/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = -1 ENOENT (No such file or directory)
access("/tmp/.mount_Text_Ef78DLY/usr/share/gnome/glib-2.0/schemas/gschemas.compiled", F_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/gnome/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = -1 ENOENT (No such file or directory)
access("/tmp/.mount_Text_Ef78DLY/usr/share/ubuntu/glib-2.0/schemas/gschemas.compiled", F_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/ubuntu/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_Text_Ef78DLY/usr/share/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = 11
fstat(11, {st_mode=S_IFREG|0777, st_size=11464, ...}) = 0
mmap(NULL, 11464, PROT_READ, MAP_PRIVATE, 11, 0) = 0x7f300a0b5000
close(11)                               = 0
access("/tmp/.mount_Text_Ef78DLY/home/me/.local/share/glib-2.0/schemas/gschemas.compiled", F_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/me/.local/share/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = -1 ENOENT (No such file or directory)
getcwd("/home/me", 4072)                = 9
access("/tmp/.mount_Text_Ef78DLY/home/me/gschemas.compiled", F_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "gschemas.compiled", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_Text_Ef78DLY/app/share/runtime-schemas/gschemas.compiled", O_RDONLY) = 11
fstat(11, {st_mode=S_IFREG|0777, st_size=42808, ...}) = 0
mmap(NULL, 42808, PROT_READ, MAP_PRIVATE, 11, 0) = 0x7f300a0aa000
close(11)                               = 0
openat(AT_FDCWD, "/tmp/.mount_Text_Ef78DLY/app/share/glib-2.0/schemas/gschemas.compiled", O_RDONLY) = 11
fstat(11, {st_mode=S_IFREG|0777, st_size=11464, ...}) = 0
mmap(NULL, 11464, PROT_READ, MAP_PRIVATE, 11, 0) = 0x7f300a0a7000
close(11)                               = 0
futex(0x7f300c7caf38, FUTEX_WAKE_PRIVATE, 2147483647) = 0
ioctl(2, TCGETS, {B38400 opost isig icanon echo ...}) = 0
getpid()                                = 8015
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=2335, ...}) = 0
write(2, "\n(gedit:8015): GLib-GIO-\33[1;31mE"..., 134
(gedit:8015): GLib-GIO-ERROR **: 21:43:42.613: Settings schema 'com.canonical.unity-gtk-module' is not installed
) = 134
--- SIGTRAP {si_signo=SIGTRAP, si_code=SI_KERNEL} ---
+++ killed by SIGTRAP +++
Trace/breakpoint trap
```

When I `unset GTK_MODULES` and then run it, _it works_! Magic. Theming is comlpetely broken though; I think we need to ship the theme along.

![](https://user-images.githubusercontent.com/2480569/77005957-96ba0280-6959-11ea-8431-3f9ced23c616.png)

**Comment by probonopd** [comment]
I get the same results with the latest continuous build, 7a8c9b3.

**Comment by sudo-give-me-coffee** [comment]
So at end is Ubuntu forcing loading modules... a rpath really not work. I added `unset GTK_MODULES` https://github.com/sudo-give-me-coffee/flatpak2appdir/commit/b6b33c4c55d48b4888dbcd9ae657a13da5aa770d, now i think that's will work seamless

> Theming is comlpetely broken though; I think we need to ship the theme along.

If we do it, the desktop integration will disappear and as it affects all GTK apps created for newer distributions (it happens with flatpaks and snaps too) and custom themes fix that, I think it is better not to send themes

**Comment by probonopd** [comment]
> If we do it, the desktop integration will disappear and as it affects all GTK apps created for newer distributions (it happens with flatpaks and snaps too) and custom themes fix that, I think it is better not to send themes

I am mixed about this. elementary OS apps, for example, don't look good when not elementary themed. Same is increasingly true for Gnome apps.

**Comment by probonopd** [comment]
Can confirm that the latest continuous build now runs on Ubuntu 16.04. (Theming broken though.)

**Comment by sudo-give-me-coffee** [comment]
> Can confirm that the latest continuous build now runs on Ubuntu 16.04.

I will close now, thank you very much for helping

> (Theming broken though.)

About themes, the user actually need to change (or update) and is relatively easy to do, this is a screenshot on Ubuntu 16.04:

![org gnome gedit](https://user-images.githubusercontent.com/20757825/77018828-b9a6df80-6976-11ea-87b4-7f34ece7d645.png)

Even native applications of Ubuntu 16.04 has the same issue

**Comment by sudo-give-me-coffee** [comment]
However adding
```
export GTK_THEME="$(mktemp -u)"
```
Forces application to use Adwaita theme
